Java/core/src/main/java/com/amazon/randomcutforest/tree/AbstractRandomCutTree.java

```diff
@@ -270,75 +270,99 @@ void updateAncestorNodesAfterDelete(NodeReference nodeReference, Point point) {
+        switchLeafReference(nodeReference, newRef);
+    }
+
+    public int getCopies(PointReference reference, boolean flagError) {
+        checkNotNull(reference, " cannot be null ");
+        NodeReference nodeReference = findLeaf(getPointFromPointReference(reference), false);
+        checkState(!flagError || getPointReference(nodeReference) == reference, "error");
+        return getMass(nodeReference);
     }
 
     /**
-     * This function deletes the point from the tree recursively. We traverse the
-     * tree based on the cut stored in each interior node until we reach a leaf
-     * node. We then delete the leaf node if the mass of the node is 1, otherwise we
-     * reduce the mass by 1. The bounding boxes continue
-     *
-     * @param nodeReference  node that we are visiting in the tree.
-     * @param point          the point that is being deleted from the tree.
-     * @param sequenceNumber the unique id of the point (if sequence enabled)
-     * @param level          the level (i.e., the length of the path to the root) of
-     *                       the node being evaluated.
+     * deletes a point from the tree
+     * 
+     * @param pointReference the reference of the point
+     * @param sequenceNumber the sequence number (in case we are storing that in the
+     *                       leaves)
+     * @return the reference used by the leaf node (after verifying equality)
      */
 
-    private void deletePoint(NodeReference nodeReference, Point point, long sequenceNumber, int level) {
+    @Override
+    public PointReference deletePoint(PointReference pointReference, long sequenceNumber) {
+        checkState(root != null, "root must not be null");
 
-        if (isLeaf(nodeReference)) {
-            Point oldPoint = getPointFromLeafNode(nodeReference);
-            if (!equals(oldPoint, point)) {
-                throw new IllegalStateException(toString(point) + " " + toString(getPointFromLeafNode(nodeReference))
-                        + " " + nodeReference + " node " + false + " Inconsistency in trees in delete step here.");
-            }
-            if (enableSequenceIndices) {
-                deleteSequenceIndex(nodeReference, sequenceNumber);
-            }
-            // decrease mass for the delete
-            if (decrementMass(nodeReference) > 0) {
-                updateAncestorPointSum(nodeReference);
-                return;
-            }
+        Point point = getPointFromPointReference(pointReference);
+        NodeReference nodeReference = findLeafAndVerify(point, true);
+        if (enableSequenceIndices) {
+            deleteSequenceIndex(nodeReference, sequenceNumber);
+        }
+        PointReference returnVal = getPointReference(nodeReference);
 
-            NodeReference parent = getParent(nodeReference);
+        // decrease mass for the delete
+        if (decrementMass(nodeReference) > 0) {
```

Comment:
Isn't the mass being decremented in `findLeafAndVerify` now? Given that someone can try to delete a point that doesn't exist, I think it makes sense to do the mass adjustment after finding the point (which I think is what we had before).

Reply:
In the current code, we should never try to delete a point which is not in the tree. But in future, we may want to. Changed.